File: integration-tests/src/test/java/dev/capylang/test/DictCollectionTest.java
package dev.capylang.test;

import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;

class DictCollectionTest {
    private static final Map<String, Integer> EXPECTED = expected();

    private static Map<String, Integer> expected() {
        var map = new LinkedHashMap<String, Integer>();
        map.put("one", 1);
        map.put("two", 2);
        map.put("three", 3);
        return map;
    }

    @Test
    void staticDict() {
        assertThat(DictCollection.staticDict()).isEqualTo(EXPECTED);
    }

    @Test
    void emptyStaticDict() {
        assertThat(DictCollection.emptyStaticDict()).isEmpty();
    }

    @Test
    void emptyStaticDictWithSpaces() {
        assertThat(DictCollection.emptyStaticDictWithSpaces()).isEmpty();
    }

    @Test
    void staticDictWithTrailingComma() {
        assertThat(DictCollection.staticDictWithTrailingComma()).isEqualTo(EXPECTED);
    }

    @Test
    void append() {
        var input = new HashMap<>(EXPECTED);

        var result = DictCollection.append(input);

        assertThat(result).isEqualTo(Map.of(
                "one", 1,
                "two", 2,
                "three", 3,
                "four", 4
        ));
        assertThat(input).isEqualTo(EXPECTED);
    }

    @Test
    void appendDict() {
        var left = new HashMap<>(Map.of("one", 1, "two", 2));
        var right = new HashMap<>(Map.of("three", 3, "four", 4));

        var result = DictCollection.appendDict(left, right);

        assertThat(result).isEqualTo(Map.of(
                "one", 1,
                "two", 2,
                "three", 3,
                "four", 4
        ));
        assertThat(left).isEqualTo(Map.of("one", 1, "two", 2));
        assertThat(right).isEqualTo(Map.of("three", 3, "four", 4));
    }

    @Test
    void appendTuple() {
        var input = new HashMap<>(EXPECTED);

        var result = DictCollection.appendTuple(input);

        assertThat(result).isEqualTo(Map.of(
                "one", 1,
                "two", 2,
                "three", 3,
                "leet", 1337
        ));
        assertThat(input).isEqualTo(EXPECTED);
    }

    @Test
    void appendTupleOverride() {
        var input = new HashMap<>(EXPECTED);

        var result = DictCollection.appendTupleOverride(input);

        assertThat(result).isEqualTo(Map.of(
                "one", 1,
                "two", 200,
                "three", 3
        ));
        assertThat(input).isEqualTo(EXPECTED);
    }

    @Test
    void remove() {
        var input = new HashMap<>(EXPECTED);

        var result = DictCollection.remove(input);

        assertThat(result).isEqualTo(Map.of(
                "one", 1,
                "three", 3
        ));
        assertThat(input).isEqualTo(EXPECTED);
    }

    @Test
    void removeDict() {
        var left = new HashMap<>(Map.of("one", 1, "two", 2, "three", 3));
        var right = new HashMap<>(Map.of("two", 1000, "three", 2000));

        var result = DictCollection.removeDict(left, right);

        assertThat(result).isEqualTo(Map.of("one", 1));
        assertThat(left).isEqualTo(Map.of("one", 1, "two", 2, "three", 3));
        assertThat(right).isEqualTo(Map.of("two", 1000, "three", 2000));
    }

    @Test
    void contains() {
        assertThat(DictCollection.contains(EXPECTED, "two")).isTrue();
        assertThat(DictCollection.contains(EXPECTED, "missing")).isFalse();
    }

    @Test
    void isEmpty() {
        assertThat(DictCollection.isEmpty(Map.of())).isTrue();
        assertThat(DictCollection.isEmpty(EXPECTED)).isFalse();
    }

    @Test
    void notIsEmpty() {
        assertThat(DictCollection.notIsEmpty(Map.of())).isFalse();
        assertThat(DictCollection.notIsEmpty(EXPECTED)).isTrue();
    }

    @Test
    void notNestedIsEmpty() {
        assertThat(DictCollection.notNestedIsEmpty(new DictCollection.DictHolder(Map.of()))).isFalse();
        assertThat(DictCollection.notNestedIsEmpty(new DictCollection.DictHolder(EXPECTED))).isTrue();
    }

    @Test
    void size() {
        assertThat(DictCollection.size(EXPECTED)).isEqualTo(3);
        assertThat(DictCollection.size(Map.of())).isEqualTo(0);
    }

    @Test
    void dictToList() {
        assertThat(DictCollection.dictToList(EXPECTED))
                .containsExactlyInAnyOrder(
                        new DictCollection.Entry("one", 1),
                        new DictCollection.Entry("two", 2),
                        new DictCollection.Entry("three", 3)
                );
    }

    @Test
    void dictToString() {
        var result = DictCollection.dictToString(EXPECTED);
        assertThat(result).isEqualTo("one 1, two 2, three 3");
    }
}

File: capy/src/e2e-cfun/js/dev/capylang/test/collection/DictCollection.test.js
const test = require('node:test');
const { assert, assertCapyEqual, generatedModule } = require('../_support');

test('DictCollection', () => {
    const dict = generatedModule('dev/capylang/test/collection/DictCollection.js');
    const expectedDict = new Map([['one', 1], ['two', 2], ['three', 3]]);
    assertCapyEqual(dict.staticDict(), expectedDict);
    assertCapyEqual(dict.emptyStaticDict(), new Map());
    assertCapyEqual(dict.append(new Map(expectedDict)), new Map([['one', 1], ['two', 2], ['three', 3], ['four', 4]]));
    assertCapyEqual(dict.appendDict(new Map([['one', 1], ['two', 2]]), new Map([['three', 3], ['four', 4]])), new Map([['one', 1], ['two', 2], ['three', 3], ['four', 4]]));
    assertCapyEqual(dict.remove(new Map(expectedDict)), new Map([['one', 1], ['three', 3]]));
    assert.equal(dict.contains(expectedDict, 'two'), true);
    assert.equal(dict.anyValue(expectedDict), true);
    assert.equal(dict.anyEntry(new Map([['two', 3]])), false);
    assert.equal(dict.allValues(new Map([['zero', 0], ['one', 1]])), false);
    assert.equal(dict.allEntries(expectedDict), true);
    assert.equal(dict.notNestedIsEmpty(new dict.DictHolder({ values: expectedDict })), true);
    assert.equal(dict.letNamedDict(), 2);
    assertCapyEqual(dict.dictToList(expectedDict), [
        new dict.Entry({ k: 'one', v: 1 }),
        new dict.Entry({ k: 'two', v: 2 }),
        new dict.Entry({ k: 'three', v: 3 }),
    ]);
    assert.equal(dict.dictToString(expectedDict), 'one 1, two 2, three 3');

    const values = new Map([['a', 1], ['b', 2]]);
    assert.equal(dict.size(values), 2);
    assert.equal(dict.appendTuple(values).get('leet'), 1337);
    assert.equal(dict.appendTupleOverride(new Map(expectedDict)).get('two'), 200);
});
